File: src-tauri/swift/focus_capture.swift
import Foundation
import AppKit
import CoreGraphics

// Usage: yoyo-focus <output-path>
// Captures a region around the mouse cursor and saves as PNG.
// Outputs JSON: {"cursor_x": f64, "cursor_y": f64, "width": f64, "height": f64}

let cropWidth: CGFloat = 800
let cropHeight: CGFloat = 600

guard CommandLine.arguments.count > 1 else {
    let err: [String: Any] = ["error": "Usage: yoyo-focus <output-path>"]
    if let data = try? JSONSerialization.data(withJSONObject: err),
       let str = String(data: data, encoding: .utf8) {
        print(str)
    }
    exit(1)
}

let outputPath = CommandLine.arguments[1]

// Get mouse cursor position (screen coordinates, top-left origin for CGEvent)
let mouseLocation = CGEvent(source: nil)?.location ?? CGPoint.zero

// Get main display bounds for clamping
let displayID = CGMainDisplayID()
let screenWidth = CGFloat(CGDisplayPixelsWide(displayID))
let screenHeight = CGFloat(CGDisplayPixelsHigh(displayID))

// On Retina displays, CGDisplayPixelsWide returns physical pixels.
// CGEvent locations and CGWindowListCreateImage rects use points.
// We need the display bounds in points.
let screenBounds = CGDisplayBounds(displayID)
let pointWidth = screenBounds.width
let pointHeight = screenBounds.height

// Calculate crop rect centered on cursor, clamped to screen bounds
var cropX = mouseLocation.x - cropWidth / 2
var cropY = mouseLocation.y - cropHeight / 2

// Clamp to screen edges
cropX = max(0, min(cropX, pointWidth - cropWidth))
cropY = max(0, min(cropY, pointHeight - cropHeight))

let cropRect = CGRect(x: cropX, y: cropY, width: cropWidth, height: cropHeight)

// Capture the screen region
guard let image = CGWindowListCreateImage(
    cropRect,
    .optionOnScreenOnly,
    kCGNullWindowID,
    .bestResolution
) else {
    let err: [String: Any] = ["error": "Failed to capture screen region"]
    if let data = try? JSONSerialization.data(withJSONObject: err),
       let str = String(data: data, encoding: .utf8) {
        print(str)
    }
    exit(1)
}

// Save as PNG
let url = URL(fileURLWithPath: outputPath)
guard let destination = CGImageDestinationCreateWithURL(url as CFURL, kUTTypePNG, 1, nil) else {
    let err: [String: Any] = ["error": "Failed to create image destination"]
    if let data = try? JSONSerialization.data(withJSONObject: err),
       let str = String(data: data, encoding: .utf8) {
        print(str)
    }
    exit(1)
}
CGImageDestinationAddImage(destination, image, nil)
guard CGImageDestinationFinalize(destination) else {
    let err: [String: Any] = ["error": "Failed to write PNG"]
    if let data = try? JSONSerialization.data(withJSONObject: err),
       let str = String(data: data, encoding: .utf8) {
        print(str)
    }
    exit(1)
}

// Output result as JSON
let result: [String: Any] = [
    "cursor_x": mouseLocation.x,
    "cursor_y": mouseLocation.y,
    "width": cropWidth,
    "height": cropHeight
]

if let jsonData = try? JSONSerialization.data(withJSONObject: result, options: []),
   let jsonString = String(data: jsonData, encoding: .utf8) {
    print(jsonString)
} else {
    print("{\"error\":\"JSON serialization failed\"}")
    exit(1)
}

File: src-tauri/src/focus_capture.rs
use serde::Deserialize;
use std::path::PathBuf;

#[derive(Debug, Deserialize)]
struct FocusCaptureOutput {
    cursor_x: f64,
    cursor_y: f64,
    width: f64,
    height: f64,
    #[serde(default)]
    error: Option<String>,
}

#[allow(dead_code)]
pub struct FocusCaptureResult {
    pub cursor_x: f64,
    pub cursor_y: f64,
    pub width: f64,
    pub height: f64,
    pub image_path: PathBuf,
}

/// Capture an 800x600 region around the mouse cursor using the compiled Swift helper.
/// Returns the cursor position, crop dimensions, and path to the saved PNG.
pub fn capture_focus_area() -> Result<FocusCaptureResult, String> {
    let focus_binary = env!("YOYO_FOCUS_BINARY");
    let output_path = std::env::temp_dir().join("yoyo-focus.png");
    let output_path_str = output_path
        .to_str()
        .ok_or("Invalid temp path for focus capture")?;

    let output = std::process::Command::new(focus_binary)
        .arg(output_path_str)
        .output()
        .map_err(|e| format!("Failed to run focus capture helper: {}", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("Focus capture helper failed: {}", stderr));
    }

    let json_str = String::from_utf8_lossy(&output.stdout);
    let parsed: FocusCaptureOutput = serde_json::from_str(&json_str).map_err(|e| {
        format!(
            "Failed to parse focus capture output: {}. Raw: {}",
            e, json_str
        )
    })?;

    if let Some(ref err) = parsed.error {
        return Err(format!("Focus capture error: {}", err));
    }

    Ok(FocusCaptureResult {
        cursor_x: parsed.cursor_x,
        cursor_y: parsed.cursor_y,
        width: parsed.width,
        height: parsed.height,
        image_path: output_path,
    })
}
